Connection: close is case insensitive

njsmith opened the pull request:
test_chunked_te was failing for me because httpbin was sending a
header like

    connection: Close

and then asks was getting confused because it was looking for "close",
not "Close". However, these are equivalent, because "Connection
options are case-insensitive" -- RFC 7230 section 6.1.

Zaab1t commented:
Wondering if `str.casefold` is better than `str.lower` in any case.

njsmith commented:
This is a `bytes` object and the spec wants ascii case insensitivity, which is what `bytes.lower` gets us. And `bytes.casefold` doesn't exist :-)

Zaab1t commented:
Gotcha. A grep suggests that you fixed them all, so I say merge!

Is this happening elsewhere? I'm not too familiar with the spec, but what about https://github.com/theelous3/asks/blob/master/asks/request_object.py#L556?
`if resp_data['headers']['transfer-encoding'] == 'chunked'`

Zaab1t commented:
Grepped for all comparisons - here's what might be relevant:

```
./auth.py:109:                if auth_dict['nonce'] == self.nonce:
./auth.py:152:                if auth_dict['nonce'] == self.nonce:
./req_structs.py:18:                        if i.host == host_loc)
./req_structs.py:29:            if i.host == host_loc:
./req_structs.py:57:        cid['aCCEPT'] == 'application/json'  # True
./req_structs.py:58:        list(cid) == ['Accept']  # True
./request_object.py:210:            if not (self.scheme == self.initial_scheme and
./request_object.py:211:               self.netloc == self.initial_netloc):
./request_object.py:332:                if l_scheme != self.scheme or l_netloc != self.netloc:
./request_object.py:540:            if header[0] == b'set-cookie':
./request_object.py:556:                if resp_data['headers']['transfer-encoding'] == 'chunked':
./response_objects.py:79:        if resp_encoding == 'gzip':
./response_objects.py:81:        elif resp_encoding == 'deflate':
```

theelous3 commented:
This is very odd. Does it actually solve the issue? The headers are stored in a case-insensitive dict.

njsmith commented:
It did switch the tests from failing to passing for me locally, yeah.
Usually case insensitive dict means that the keys are case insensitive, not
the values?

On Sep 19, 2017 1:50 AM, "M.J." <notifications@github.com> wrote:

> Merged #40 <https://github.com/theelous3/asks/pull/40>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/theelous3/asks/pull/40#event-1254989218>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAlOaKqtG4phFhTpUU53C_p-lveMZViYks5sj4BygaJpZM4Pb5DR>
> .
>

theelous3 commented:
Yep, just had the auld brain turned off for a few minutes there. 

It's failing on pypy atm on travis but that's not the end of the world. I'll take a look at it over lunch.

njsmith commented:
The pypy issue is a bug in `regex` that will hopefully be fixed soon: https://bitbucket.org/mrabarnett/mrab-regex/issues/253/run-into-error-under-pypy-580

theelous3 commented:
Yep. I figured it best to merge / update pypi anyway for sake of the majority, even if it means the build appears to fail.

njsmith commented:
So apparently the `regex` bug is now fixed, which then revealed a bug in pypy, which is also [now fixed](https://bitbucket.org/pypy/pypy/issues/2663/regex-package-on-python-3-uses), so hopefully everything will be working once PyPy3 5.9 comes out (should be in the next few days).
